Cell wise grain size particle update

gassmoeller opened the pull request:
@jdannberg: This is a first try to play with the cell-wise particle update for grain size. I still get different test results then before, and the code is not very optimized, but it is a starting point for testing.

jdannberg commented on include/aspect/particle/property/grain_size.h:
```diff
@@ -64,10 +64,10 @@ namespace aspect
            * @copydoc aspect::Particle::Property::Interface::update_particle_property()
```
```suggestion
           * @copydoc aspect::Particle::Property::Interface::update_particle_properties()
```
?

jdannberg commented on include/aspect/particle/property/interface.h:
```diff
@@ -324,6 +324,37 @@ namespace aspect
 
+          /**
+           * Update function. This function is called every time an update is
+           * request by need_update() for every particle for every property.
```
```suggestion
           * requested by need_update() for every particle for every property.
```

jdannberg commented on include/aspect/particle/property/interface.h:
```diff
@@ -324,6 +324,37 @@ namespace aspect
           initialize_one_particle_property (const Point<dim> &position,
                                             std::vector<double> &particle_properties) const;
 
+          /**
+           * Update function. This function is called every time an update is
+           * request by need_update() for every particle for every property.
+           * It is obvious that
+           * this function is called a lot, so its code should be efficient.
+           * The interface provides a default implementation that does nothing,
+           * therefore derived plugins that do not require an update do not
+           * need to implement this function.
+           *
+           * @param [in] data_position An unsigned integer that denotes which
+           * component of the particle property vector is associated with the
+           * current property. For properties that own several components it
+           * denotes the first component of this property, all other components
+           * fill consecutive entries in the @p particle_properties vector.
+           *
+           * @param [in] solution The values of the solution variables at the
+           * current particle position.
+           *
+           * @param [in] gradients The gradients of the solution variables at
+           * the current particle position.
+           *
+           * @param [in,out] particles The particles that are updated within
+           * the call of this function.
+           */
```
Can you update the documentation to reflect that it's now more than one particle?

jdannberg commented on include/aspect/particle/property/interface.h:
```diff
@@ -565,9 +599,9 @@ namespace aspect
            * called once every time step for every particle.
```
Can you update this?

jdannberg commented on source/particle/property/grain_size.cc:
```diff
@@ -66,39 +66,49 @@ namespace aspect
+        material_inputs  = MaterialModel::MaterialModelInputs<dim>(solution.size(), this->n_compositional_fields());
+        material_outputs = MaterialModel::MaterialModelOutputs<dim>(solution.size(), this->n_compositional_fields());
+        material_inputs.requested_properties = MaterialModel::MaterialProperties::reaction_terms;
+        material_inputs.current_cell = typename DoFHandler<dim>::active_cell_iterator(*particles.begin()->get_surrounding_cell(),
```
Can it ever happen that someone passes a list of particles that are not all in one cell?

jdannberg commented on source/particle/property/grain_size.cc:
```diff
@@ -66,39 +66,49 @@ namespace aspect
 
-        material_inputs.temperature[0] = solution[this->introspection().component_indices.temperature];
+        unsigned int i = 0;
+        for (auto particle: particles)
```
Is there a reason use particle as counter rather than i?

jdannberg commented on source/particle/world.cc:
```diff
@@ -1067,15 +1065,10 @@ namespace aspect
-                typename ParticleHandler<dim>::particle_iterator_range
-                particles_in_cell = particle_handler->particles_in_cell(cell);
-
                 // Only update particles, if there are any in this cell
```
```suggestion
                // Only update particles if there are any in this cell
```

jdannberg reviewed (commented):
I like it! No idea why the tests are different though. 
And have you looked at the timing different in the grain_size_growth_one_cell_particle test? And maybe the grain size ridge test?

jdannberg commented:
I compared to the current main for the grain_size_growth_one_cell_particles test:
main: Particles: Update properties 3.92s
your branch: Particles: Update properties 0.995s

gassmoeller commented:
Nice improvements! I addressed your comments and converted one more particle property to the new interface (`composition`). The other properties can be updated in a follow-up PR.

Everything should be ready for another review.

jdannberg commented on include/aspect/particle/property/interface.h:
```diff
@@ -324,6 +324,39 @@ namespace aspect
+           * component of each particle property vector is associated with the
+           * current property. For properties that own several components it
+           * denotes the first component of this property, all other components
+           * fill consecutive entries in the @p particle_properties vector.
```
What is the @p particle_properties vector?

jdannberg commented on include/aspect/particle/property/composition.h:
I think you should just delete this whole plugin. Especially since the description does not make it very clear how this is different from initial composition, and that you should REALLY, REALLY not use this and instead use initial composition.

jdannberg reviewed (approved):
Looks good to me, other than the iteration over the particle with the index i you have to increment in the loop. But I am not sure how to fix that. 

And you still need to update the tests.

gassmoeller commented:
I have removed the `composition` plugin from this PR, fixed the documentation issue and updated the tests. I have some ideas for further optimizations, but they can wait for a follow-up PR. If the testers are happy this is good to go from my side.
